### PR TITLE
runtime-rs: Add dedicated CSI driver for DirectVolume support in Kata

### PR DESCRIPTION
**Pull request by Apokleos**
Bridge the gap between user requirements for direct block device access
  and the DirectVolume capabilities provided by Kata runtimes
  (kata-runtime/runtime-rs), and facilitate seamless integration with CSI
  to improve user experience.
  
  It aims to integrate DirectVolume CSI support into Kata, enabling users
  to benefit from its performance and flexibility advantages.
  
  Fixes: #8602
  
  Signed-off-by: alex.lyn <alex.lyn@antgroup.com>

### CODE REVIEW

File: src/tools/csi-kata-directvolume/pkg/directvolume/healthcheck.go [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+	klog.V(3).Infof("fs available: %+v", fsavailable)
+
+	return fsavailable > 0, nil
+}
+
+func (dv *directVolume) doHealthCheckInControllerSide(volID string) (bool, error) {
+	volumePath := dv.getVolumePath(volID)
+	klog.V(3).Infof("Volume with ID %s has path %s.", volID, volumePath)
+	spExist, err := utils.CheckPathExist(volumePath)
```
Comment (zjumoon01) [inline src/tools/csi-kata-directvolume/pkg/directvolume/healthcheck.go:149]: Is volumePath in controller ?  I see volumePath is checked in doHealthCheckInNodeSide below

File: src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+		errMsg := status.Errorf(codes.Internal, "setup storage for volume '%s' failed", volumeID)
+		return &csi.NodeStageVolumeResponse{}, errMsg
+	}
+
+	// /full_path_on_host/VolumeId/
+	deviceUpperPath := filepath.Dir(*devicePath)
+	if canMnt, err := utils.CanDoBindMount(dv.config.safeMounter, stagingTargetPath); err != nil {
+		return nil, err
+	} else if !canMnt {
+		klog.Infof("Cannot mount to target path: %s", stagingTargetPath)
```
Comment (zjumoon01) [inline src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go:241]: Is stagingTargetPath already mounted in this case ? The log showing "already mounted" looks better

File: src/tools/csi-kata-directvolume/cmd/directvolplugin/main.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	// Set by the build process
+	version = ""
+)
+
+func main() {
+	cfg := directvolume.Config{
+		VendorVersion: version,
+	}
+
+	flag.StringVar(&cfg.Endpoint, "endpoint", "unix://tmp/csi.sock", "CSI endpoint")
```
[the remark this answers is not in the document]
Comment (Apokleos) [reply src/tools/csi-kata-directvolume/cmd/directvolplugin/main.go]: Thx @zjumoon01 
fixed

File: src/tools/csi-kata-directvolume/pkg/directvolume/healthcheck.go (hunk 1 above)
Comment (Apokleos) [reply src/tools/csi-kata-directvolume/pkg/directvolume/healthcheck.go]: useless codes removed

File: src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+	}
+	klog.Infof("add direct volume successfully.")
+
+	var vol state.Volume
+	var err error
+	if vol, err = dv.state.GetVolumeByID(volumeId); err != nil {
+		capInt64, _ := strconv.ParseInt(req.VolumeContext[utils.CapabilityInBytes], 10, 64)
+		volName := req.VolumeContext[utils.DirectVolumeName]
+		kind := req.VolumeContext[storageKind]
+		vol, _ := dv.createVolume(volumeId, volName, capInt64, kind)
```
[the remark this answers is not in the document]
Comment (Apokleos) [reply src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go]: fixed

File: src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+	dv.config.VolumeDevices[volumeID] = *devicePath
+
+	klog.Infof("directvolume: volume %s has been staged.", stagingTargetPath)
+
+	volInStat, err := dv.state.GetVolumeByID(req.VolumeId)
+	if err != nil {
+		capInt64, _ := strconv.ParseInt(req.VolumeContext[utils.CapabilityInBytes], 10, 64)
+		volName := req.VolumeContext[utils.DirectVolumeName]
+		kind := req.VolumeContext[storageKind]
+		vol, _ := dv.createVolume(volumeID, volName, capInt64, kind)
```
[the remark this answers is not in the document]
Comment (Apokleos) [reply src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go]: fixed

File: src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go (hunk 2 above)
Comment (Apokleos) [reply src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go]: fixed

**Review by lifupan — APPROVED** [review]

File: src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+
+	if err := dv.config.safeMounter.DoBindmount(stagingTargetPath, targetPath, "", options); err != nil {
+		errMsg := fmt.Sprintf("failed to bindmount device: %s at %s: %s", stagingTargetPath, targetPath, err.Error())
+		klog.Infof("do bindmount failed: %v.", errMsg)
+		return nil, status.Error(codes.Aborted, errMsg)
+	}
+
+	// kata-containers DirectVolume add
+	mountInfo := utils.MountInfo{
+		VolumeType: "directvol",
```
Comment (Tim-0731-Hzt) [inline src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go:92]: Could you make this field configurable?

**Review by Tim-0731-Hzt — APPROVED** [review]
Thanks @Apokleos LGTM

File: src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go (hunk 6 above)
Comment (Apokleos) [reply src/tools/csi-kata-directvolume/pkg/directvolume/nodeserver.go]: Thx @Tim-0731-Hzt  fixed

**Pull request merged**